File: LowLevelDesign/java/task-scheduler/src/com/taskscheduler/observer/LoggingObserver.java
package com.taskscheduler.observer;

import com.taskscheduler.ScheduledTask;

import java.time.LocalDateTime;

public class LoggingObserver implements  TaskExecutionObserver {

    @Override
    public void onTaskStarted(ScheduledTask task) {
        System.out.printf("[LOG - %s] [%s] Task %s started.%n", LocalDateTime.now(), Thread.currentThread().getName(), task.getId());
    }

    @Override
    public void onTaskCompleted(ScheduledTask task) {
        System.out.printf("[LOG - %s] [%s] Task %s completed successfully.%n", LocalDateTime.now(), Thread.currentThread().getName(), task.getId());
    }

    @Override
    public void onTaskFailed(ScheduledTask task, Exception e) {
        System.out.printf("[LOG - %s] [%s] Task %s failed %s%n", LocalDateTime.now(), Thread.currentThread().getName(), task.getId(), e.getMessage());
    }
}

File: LowLevelDesign/task-scheduler-nodejs/src/observer/LoggingObserver.js
import { TaskExecutionObserver } from './TaskExecutionObserver.js';

/**
 * LoggingObserver - Logs task execution events
 */
export class LoggingObserver extends TaskExecutionObserver {
    onTaskStarted(task) {
        console.log(`[LOG - ${new Date().toISOString()}] Task ${task.getId()} started.`);
    }

    onTaskCompleted(task) {
        console.log(`[LOG - ${new Date().toISOString()}] Task ${task.getId()} completed successfully.`);
    }

    onTaskFailed(task, error) {
        console.log(`[LOG - ${new Date().toISOString()}] Task ${task.getId()} failed: ${error.message}`);
    }
}
